src/pdu_loop/mod.rs

```diff
@@ -76,25 +78,51 @@ impl<const MAX_FRAMES: usize, const MAX_PDU_DATA: usize> PduStorage<MAX_FRAMES,
     }
 
     /// Get a reference to this `PduStorage` with erased lifetimes.
-    pub const fn as_ref(&self) -> PduStorageRef {
+    pub const fn as_ref<'a>(&'a self) -> PduStorageRef<'a> {
+        let num_frames = unsafe { &*self.frames.get() }.len();
+
         PduStorageRef {
-            frames: &self.frames,
-            frame_data: unsafe {
-                core::slice::from_raw_parts(
-                    self.frame_data.as_ptr() as *const UnsafeCell<&[u8]>,
-                    self.frame_data.len(),
-                )
-            },
-            max_pdu_data: MAX_PDU_DATA,
+            frames: NonNull::new(self.frames.get().cast::<pdu_frame::Frame>()).unwrap(),
+            num_frames,
+            frame_data: NonNull::new(self.frame_data.get().cast::<u8>()).unwrap(),
+            frame_data_len: MAX_PDU_DATA,
+            _lifetime: PhantomData,
         }
     }
 }
 
 #[derive(Debug)]
 pub struct PduStorageRef<'a> {
-    frame_data: &'a [UnsafeCell<&'a [u8]>],
-    frames: &'a [UnsafeCell<pdu_frame::Frame>],
-    max_pdu_data: usize,
+    frames: NonNull<pdu_frame::Frame>,
+    num_frames: usize,
+    frame_data: NonNull<u8>,
+    frame_data_len: usize,
+    _lifetime: PhantomData<&'a ()>,
+}
+
+impl<'a> PduStorageRef<'a> {
+    fn frame(&self, idx: u8) -> Result<(&mut pdu_frame::Frame, &mut [u8]), Error> {
+        let idx = usize::from(idx);
+
+        if idx > self.num_frames {
+            return Err(Error::Pdu(PduError::InvalidIndex(idx)));
+        }
+
+        let frame = unsafe {
+            &mut *self
+                .frames
+                .as_ptr()
+                .add(core::mem::size_of::<pdu_frame::Frame>() * idx)
```

Comment:
```suggestion
                .add(idx)
```

`add(i)` already increments `size_of::<T>() * i`.